util/flexpipeviewer.py

```diff
@@ -0,0 +1,205 @@
+    "m": "cyan",
+    "z": "green",
+    "c": "blue",
+    "s": "magenta"
+}
+
+# f: creationTick (inst created)
+# d: decodeTick (inst decoded)
+# i: issueTick (issued)
+# e: beginExecuteTick (execute request sent)
+# a: effectiveAddressTick (address available)
+# m: beginMemoryTick (memory access request sent)
+# z: completionTick (execute stage is done)
+# c: commitTick (inst committed)
+# s: squashTick (inst squashed)
+
+stages = ["f", "d", "i", "e", "a", "m", "z", "c", "s"]
+
+stages_with_color = [bg_color_map[stage_color_map[stage]] + stage +
+                    bg_color_map["black"] for stage in stages]
+
+bg_color_char_lambda = lambda color_char : bg_color_map[color_char[0]] \
+                       + color_char[1] + bg_color_map["black"]
+
+_stage_color_map = {k: v for k, v in stage_color_map.items()}
+
+def colorize_pipeline(pipeline, start_pos, end_pos, bg_color_lambda):
+    i = start_pos
+    pipewidth = len(pipeline)
+    pipecolor = [("black", pipechar) for pipechar in pipeline]
+    curr_color = stage_color_map[pipeline[i]]
+    while not i == end_pos:
+        _stage_color_map["."] = curr_color
+        _stage_color_map["="] = curr_color
+        pipecolor[i] = (_stage_color_map[pipeline[i]], pipeline[i])
+        curr_color = pipecolor[i][0]
+        i = (i + 1) % pipewidth
+    return list(map(bg_color_char_lambda, pipecolor))
+
+def no_colorize_pipeline(pipeline, start_pos, end_pos, bg_color_lambda):
+    return pipeline
+
+def parser(input_file, output_file, tick_start, tick_end, color, width,
+           cycle_time):
+    curr_batch_seq_num = 0
+    insts = {}
+    dep = {}
+    max_seq_num = -1
+    width_ticks = cycle_time * width
+
+    if color:
+        coloring_fnct = colorize_pipeline
+    else:
+        coloring_fnct = no_colorize_pipeline
+
+    with open(input_file, "r") as f:
+        for line in f:
+            line = line.strip().split(";")
+            if line[0] == "pipe":
+                # decode the instruction info
+                seq_num = int(line[1])
+                pc = int(line[2], 16)
+                upc = int(line[3], 16)
+                inst = line[4].strip()
+                creationTick = int(line[5])
+                decodeTick = int(line[6])
+                issueTick = int(line[7])
+                beginExecuteTick = int(line[8])
+                effAddredTick = int(line[9])
+                beginMemoryTick = int(line[10])
+                completionTick = int(line[11])
+                commitTick = int(line[12])
+                squashTick = int(line[13])
+
+
+                # turn the info into viewable pipeline
+                if squashTick == NEG_ONE:
+                    pipe = ["."] * (width)
+                else:
+                    pipe = ["="] * (width)
+                curr_tick = creationTick - creationTick % width_ticks
+                pipe[(creationTick - curr_tick) // cycle_time] = "c"
+                for tick, name in zip([creationTick, decodeTick, issueTick,
```

Comment:
I think this can be written cleaner. For instance:

```
    data = [creationTick, decodeTick, issueTick, beginExecuteTick, effAddredTick,
    beginMemoryTick, completionTick, commitTick, squashTick]
    for tick, name in zip(data, stages):
        ...
```

Reply:
I think I'll use a tuple or a struct to do this as the ticks should be immutable anyway. I agree that doing this way will make the code cleaner.